src/CommunityToolkit.HighPerformance/Extensions/StreamExtensions.cs

```diff
@@ -207,27 +207,45 @@ public static unsafe T Read<T>(this Stream stream)
     {
 #if NETSTANDARD2_1_OR_GREATER
         T result = default;
-        int length = sizeof(T);
+        int bytesOffset = 0;
 
-        unsafe
+        // As per Stream.Read's documentation:
+        // "The total number of bytes read into the buffer. This can be less than the number of bytes allocated in the
+        // buffer if that many bytes are not currently available, or zero (0) if the end of the stream has been reached."
+        // Because of this, we have to loop until all requires bytes have been read, and only throw if the return is 0.
+        do
         {
-            if (stream.Read(new Span<byte>(&result, length)) != length)
+            int bytesRead = stream.Read(new Span<byte>((byte*)&result + bytesOffset, sizeof(T) - bytesOffset));
```

Comment:
For the .NET 7 target `Stream.ReadExactly` could be used. It boils down the be [almost a one-liner](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4B0AkgPIDchhGAzOTkuVqQMKkDehUsPJsKnAK4A7AM4BDAGYxSAFVIAlGHIAmAHhUA+ABQAXABYBLGaQDKJ2HIC2pGfa2OAlKQDuZmLFVSEFJpRzkpOQBzGG0hEUECESTA2BkJABsTUgBeUm0YBTkMk0ZE5OFXB0dKTR0AUQQ5MBN0gE8jKRhvWwAHcN1gVpMYYyNB4YAqDwAyVOKAGhcLAC8YCAUjFQ9t0vLhDAB2UjnM3ZEAX0JzoA).